walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,169 @@
+     * time range for. Returns an empty list.
+     */
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return new ArrayList<TimeRange>();
+    }
+    
+    ArrayList<Event> eventsArr = sortEventsByStart(events);
+
+    if (request.hasOptionalAttendees()) {
+      ArrayList<TimeRange> optionalSlots = findTimeSlots(eventsArr, request, 
+        /* includeOptionalAttendees = */ true);
+      if (optionalSlots.isEmpty() && request.hasAttendees()) {
+        return findTimeSlots(eventsArr, request, /* includeOptionalAttendees = */ false);
+      }
+      return optionalSlots;
+    } else {
+      return findTimeSlots(eventsArr, request, /* includeOptionalAttendees = */ false);
+    }
+  }
+
+  /*
+   * Sort events by the start of their TimeRanges.
+   */
+  private ArrayList<Event> sortEventsByStart(Collection<Event> events) {
+    ArrayList<Event> eventsArr = new ArrayList<>(events);
+    Collections.sort(eventsArr, new Comparator<Event>() {
+      @Override
+      public int compare(Event a, Event b) {
+        return Long.compare(a.getWhen().start(), b.getWhen().start());
+      }
+    });
+    return eventsArr;
+  }
+
+  /*
+   * Finds all the possible time ranges throughout the day that the requested
+   * meeting can take place, given already-scheduled events.
+   * 
+   * @param events: List of already-scheduled events that may conflict with the times
+   * in which the request can be scheduled.
+   * @param request: Requested meeting to be scheduled into the day.
+   * @param includeOptionalAttendees: Find time ranges that work for both required
+   * and optional attendees (as opposed to only required attendees).
+   * @return List of all possible time ranges that the meeting request can take place within.
+   */
+  private ArrayList<TimeRange> findTimeSlots(Collection<Event> events, MeetingRequest request,
+    boolean includeOptionalAttendees) {
+    /*
+     * Slots stores the possible time ranges for the meeting request
+     * to take place. By default, this stores the entire day, and is gradually
+     * cut down into fragments if conflicts arise with events.
+     */
+    ArrayList<TimeRange> slots = new ArrayList<TimeRange>();
+    slots.add(TimeRange.WHOLE_DAY);
+
+    /*
+     * A copy of slots that stores new changes to slots without interferring
+     * with the for loop below.
+     */
+    ArrayList<TimeRange> curSlots = new ArrayList<>(slots);
+
+    for (Event event : events) {
+      TimeRange eventTime = event.getWhen();
+      slots = curSlots;
+
+      /*
+       * Finds whether there is any overlap between event's attendees and request's attendees.
+       * If there is no overlap, then this event is skipped.
+       */
+      if (hasNothingInCommon(event.getAttendees(), request.getAttendees())) {
+        if (!includeOptionalAttendees || 
+          hasNothingInCommon(event.getAttendees(), request.getOptionalAttendees())) {
+          continue;
+        }
+      }
```

Comment:
The spacing in this suggestion may be a little off, but this can be combined into one:
```suggestion
      if ((hasNothingInCommon(event.getAttendees(), request.getAttendees()))) &&
             (!includeOptionalAttendees ||  
             hasNothingInCommon(event.getAttendees(), request.getOptionalAttendees()))){
          continue;
      }
```

Also, I love this change!  I think it's much easier to read now.